csharp/ql/lib/semmle/code/csharp/security/auth/ActionMethods.qll

```diff
@@ -0,0 +1,107 @@
+/** Common definitions for queries checking for access control measures on action methods. */
+
+import csharp
+import semmle.code.csharp.frameworks.microsoft.AspNetCore
+import semmle.code.csharp.frameworks.system.web.UI
+
+/** A method representing an action for a web endpoint. */
+abstract class ActionMethod extends Method {
+  /**
+   * Gets a string that can indicate what this method does to determine if it should have an auth check;
+   * such as its method name, class name, or file path.
+   */
+  string getADescription() {
+    result = [this.getName(), this.getDeclaringType().getBaseClass*().getName(), this.getARoute()]
+  }
+
+  /** Holds if this method may represent a stateful action such as editing or deleting */
+  predicate isEdit() {
+    exists(string str |
+      str =
+        this.getADescription()
+            // separate camelCase words
+            .regexpReplaceAll("([a-z])([A-Z])", "$1_$2")
+            .toLowerCase() and
+      str.regexpMatch(".*(edit|delete|modify|change).*") and
```

Comment:
I am wondering if it would be possible to come up with a better strategy for detecting such stateful actions. Obviously the current approach is prone to miss things that are not in the list or where code bases use non-English names for things. Could we check for data / taint flow to known sinks like database access etc. instead? That might still not catch everything, but might be more accurate / inclusive of global customers than just looking at names.

Reply:
I did try a dataflow based approach to reduce dependence on method names, however I ran into issues (written up [here](https://github.com/github/codeql-csharp-team/issues/358#issuecomment-1650377114)) with it not working for the case in webgoat; essentially due to a framework-internal control flow step that would need to be modeled specially.

Reply:
I don't think it is unreasonable to do some framework-specific modelling for common web frameworks. I wonder if @michaelnebel or @hvitved have any thoughts on how this particular data flow issue with WebForms could be modelled well?

Reply:
Since as you mention in another comment, WebForms is outdated and not commonly used, does it make sense to implement specific flow steps for it?